src/Plugin/Field/FieldFormatter/AvPortalPhotoResponsiveFormatter.php

```diff
@@ -0,0 +1,348 @@
+  public static function isApplicable(FieldDefinitionInterface $field_definition): bool {
+    if ($field_definition->getTargetEntityTypeId() !== 'media') {
+      return FALSE;
+    }
+
+    if (!parent::isApplicable($field_definition)) {
+      return FALSE;
+    }
+
+    // We don't want this formatter used if the Responsive Image module is not
+    // installed.
+    if (!\Drupal::moduleHandler()->moduleExists('responsive_image')) {
+      return FALSE;
+    }
+
+    // We don't want to use this formatter if there are no responsive image
+    // styles on the site.
+    if (empty(\Drupal::entityTypeManager()->getStorage('responsive_image_style')->loadMultiple())) {
+      return FALSE;
+    }
+
+    $media_type_id = $field_definition->getTargetBundle();
+    if (!$media_type_id) {
+      // We need to allow to use this formatter also in cases where the field is
+      // not bundle-specific, like when it's a base field or it's used as a
+      // formatter for Views. So we rely on developers not using this formatter
+      // on fields where it doesn't make any sense.
+      return TRUE;
+    }
+
+    $media_type = MediaType::load($media_type_id);
+    return $media_type && $media_type->getSource() instanceof MediaAvPortalPhotoSource;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode): array {
+    $elements = [];
+
+    foreach ($items as $delta => $item) {
+      $elements[$delta] = $this->viewElement($item);
+    }
+
+    $style_id = $this->getSetting('responsive_image_style');
+    if (!$style_id) {
+      return $elements;
+    }
+
+    $responsive_image_style = $this->entityTypeManager->getStorage('responsive_image_style')->load($style_id);
+    if (!$responsive_image_style instanceof ResponsiveImageStyleInterface) {
+      return $elements;
+    }
+
+    $cache = new CacheableMetadata();
+    $cache->addCacheableDependency($responsive_image_style);
+    if ($responsive_image_style->hasImageStyleMappings()) {
+      $image_styles = $responsive_image_style->getImageStyleIds();
+      $image_styles = $this->entityTypeManager->getStorage('image_style')->loadMultiple($image_styles);
+      foreach ($image_styles as $image_style) {
+        $cache->addCacheableDependency($image_style);
+      }
+    }
+
+    $cache->applyTo($elements);
+
+    return $elements;
+  }
+
+  /**
+   * Renders a single field item.
+   *
+   * @param \Drupal\Core\Field\FieldItemInterface $item
+   *   The individual field item.
+   *
+   * @return array
+   *   The Drupal element.
+   */
+  protected function viewElement(FieldItemInterface $item): array {
+    $main_property = $item->getFieldDefinition()
+      ->getFieldStorageDefinition()
+      ->getMainPropertyName();
+
+    $resource_ref = $item->{$main_property};
+
+    if (empty($resource_ref)) {
+      return [];
+    }
+
+    $resource = $this->avPortalClient->getResource($resource_ref);
+
+    if (!$resource instanceof AvPortalResource) {
```

Comment:
What do we check again that this assets accessible on remote service? Because we already did this as I understand on step of adding media entity.

Reply:
I think this is because we store the reference to an AvPortal resource, not the resource itself, so when we want to render it it may not be available.

Reply:
Exactly. We only store the reference.